protocols/v2/roles-logic-sv2/src/utils.rs

```diff
@@ -155,10 +158,126 @@ impl<T> Mutex<T> {
     }
 }
 
+#[derive(Clone)]
+pub struct Coinbase {
+    pub tx: Transaction,
+    pub script_sig_prefix_len: usize,
+}
+
+impl Coinbase {
+    pub fn new(
+        script_sig_prefix: Vec<u8>,
+        version: i32,
+        lock_time: u32,
+        sequence: u32,
+        coinbase_outputs: Vec<TxOut>,
+        additional_coinbase_script_data_len: usize,
+        extranonce_len: u8,
+    ) -> Self {
+        let mut script_sig = script_sig_prefix.clone();
+        script_sig.extend_from_slice(&vec![0_u8; additional_coinbase_script_data_len]);
+        script_sig.extend_from_slice(&vec![0; extranonce_len as usize]);
+        let tx_in = TxIn {
+            previous_output: OutPoint::null(),
+            script_sig: script_sig.into(),
+            sequence: bitcoin::Sequence(sequence),
+            witness: Witness::from(vec![] as Vec<Vec<u8>>), /* empty witness (for bip141-stripped
+                                                             * coinbase_tx_prefix and
+                                                             * coinbase_tx_suffix) */
+        };
+        let tx = Transaction {
+            version: TxVersion::non_standard(version),
+            lock_time: LockTime::from_consensus(lock_time),
+            input: vec![tx_in],
+            output: coinbase_outputs.to_vec(),
+        };
+
+        Self {
+            tx,
+            // TODO: move additional_coinbase_script_data to extranonce_prefix
+            // (part of original PR #1248)
+            script_sig_prefix_len: script_sig_prefix.len() + additional_coinbase_script_data_len,
+        }
+    }
+
+    // serialize input from Transaction object
+    pub fn serialized_input(&self) -> Vec<u8> {
+        let input = self.tx.input[0].clone(); // coinbase only has one input
+        let mut serialized_input = Vec::new();
+        serialized_input.extend_from_slice(input.previous_output.txid.as_byte_array());
+        serialized_input.extend_from_slice(&input.previous_output.vout.to_le_bytes());
+        serialized_input.push(input.script_sig.len() as u8);
+        serialized_input.extend_from_slice(input.script_sig.as_bytes());
+        serialized_input.extend_from_slice(&input.sequence.0.to_le_bytes());
+        serialized_input
+    }
+
+    // serialize outputs from Transaction object
+    fn serialized_outputs(&self) -> Vec<Vec<u8>> {
+        self.tx.output.iter().map(consensus::serialize).collect()
+    }
+
+    // coinbase_tx_prefix is the LE bytes concatenation of the following coinbase fields:
+    // - tx version
+    // - input count
+    // - input outPoint
+    // - input index (vout)
+    // - scriptSig length
+    // - scriptSig prefix
+    //
+    // we skip the bip141 marker and flag (which usually come after tx version)
+    // because those are only used for computing the `wtxid`
+    // while the legacy `txid` is what is actually used for computing the merkle root
+    pub fn coinbase_tx_prefix(self) -> Result<B064K<'static>, Error> {
```

Comment:
```suggestion
    pub fn coinbase_tx_prefix(&self) -> Result<B064K<'static>, Error> {
```

Ownership is not needed here. This should allow you to remove all the `.clone()` calls for `Coinbase`.